common/membership/grpc_resolver_test.go

```diff
@@ -0,0 +1,122 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package membership_test
+
+import (
+	"context"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.temporal.io/server/common/membership"
+	"go.temporal.io/server/common/primitives"
+	"go.temporal.io/server/internal/nettest"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"net"
+	"testing"
+	"time"
+)
+
+func TestGRPCResolverBuilder_Build(t *testing.T) {
+	t.Parallel()
+
+	// There's a lot of channel stuff in this test, so we use a context to make sure we don't hang forever if something
+	// goes wrong.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	ctrl := gomock.NewController(t)
+	sr := membership.NewMockServiceResolver(ctrl)
+
+	// On the first call to [ServiceResolver.Members], return an empty list of members
+	sr.EXPECT().Members().Return([]membership.HostInfo{})
+	// Once our resolver registers a listener to membership changes, get a hold of the channel it's listening on.
+	sr.EXPECT().AddListener(gomock.Any(), gomock.Any()).Do(func(_ string, ch chan<- *membership.ChangedEvent) {
+		// After the first call to [ServiceResolver.Members] returns an empty list, expect our resolver to request a
+		// refresh of the members list. When it does, notify the listener that the members list has changed and return
+		// a single member on the next call to [ServiceResolver.Members].
+		sr.EXPECT().RequestRefresh().Do(func() {
+			sr.EXPECT().Members().Return([]membership.HostInfo{
```

Comment:
ooh, a triple-nested EXPECT!

Reply:
Made it double-nested to avoid raising too many eyebrows